trie/inspect_trie.go

```diff
@@ -0,0 +1,279 @@
+package trie
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"runtime"
+	"sort"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/rlp"
+	"github.com/olekukonko/tablewriter"
+	"golang.org/x/sync/semaphore"
+)
+
+const (
+	// defaultTriedbcacheSize default triedb cache size in hash scheme. To prevent memory leaks.
+	defaultTriedbcacheSize = 1024 * 1024 * 1024
+)
+
+type Inspector struct {
+	trie           *Trie // traverse trie
+	db             *Database
+	stateRootHash  common.Hash
+	blocknum       uint64
+	root           node // root of triedb
+	totalNum       uint64
+	wg             sync.WaitGroup
+	statLock       sync.RWMutex
+	result         map[string]*trieTreeStat
+	sem            *semaphore.Weighted
+	eoaAccountNums uint64
+}
+
+type trieTreeStat struct {
+	is_account_trie    bool
+	theNodeStatByLevel [15]nodeStat
+	totalNodeStat      nodeStat
+}
+
+type nodeStat struct {
+	ShortNodeCnt atomic.Uint64
+	FullNodeCnt  atomic.Uint64
+	ValueNodeCnt atomic.Uint64
+}
+
+func (ns *nodeStat) IsEmpty() bool {
+	if ns.FullNodeCnt.Load() == 0 && ns.ShortNodeCnt.Load() == 0 && ns.ValueNodeCnt.Load() == 0 {
+		return true
+	}
+	return false
+}
+
+func (trieStat *trieTreeStat) AtomicAdd(theNode node, height uint32) {
+	switch (theNode).(type) {
+	case *shortNode:
+		trieStat.totalNodeStat.ShortNodeCnt.Add(1)
+		trieStat.theNodeStatByLevel[height].ShortNodeCnt.Add(1)
+	case *fullNode:
+		trieStat.totalNodeStat.FullNodeCnt.Add(1)
+		trieStat.theNodeStatByLevel[height].FullNodeCnt.Add(1)
+	case valueNode:
+		trieStat.totalNodeStat.ValueNodeCnt.Add(1)
+		trieStat.theNodeStatByLevel[height].ValueNodeCnt.Add(1)
+	default:
+		panic(errors.New("invalid node type for statistics"))
+	}
+}
+
+type stringWriter struct {
+	builder strings.Builder
+}
+
+func (sw stringWriter) Write(p []byte) (n int, err error) {
+	return sw.builder.Write(p)
+}
+
+func (trieStat *trieTreeStat) Display(ownerAddress string, treeType string) string {
+	sw := stringWriter{}
```

Comment:
can't you just use a `strings.Builder` ?

Reply:
Yes, we can't. 

Cannot use 'builder' (type strings.Builder) as the type io.Writer Type does not implement 'io.Writer' as the 'Write' method has a pointer receiver

Reply:
Then you're supposed to use a pointer: use a `new(strings.Builder)`